python/ray/dataframe/dataframe.py

```diff
@@ -3636,3 +3920,92 @@ def iloc(self):
         raise NotImplementedError(
             "To contribute to Pandas on Ray, please visit "
             "github.com/ray-project/ray.")
+
+    def _copartition(self, other, new_index):
+        """Colocates the values of other with this for certain operations.
+
+        Args:
+            other: The other DataFrame to copartition with.
+
+        Returns:
+            Two new sets of partitions, copartitioned and zipped.
+        """
+        # Put in the object store so they aren't serialized each iteration.
+        old_self_index = ray.put(self.index)
+        new_index = ray.put(new_index)
+        old_other_index = ray.put(other.index)
+
+        new_partitions_self = [_reindex_helper.remote(df, old_self_index,
+                                                      new_index, 1)
+                               for df in self._col_partitions]
+
+        new_partitions_other = [_reindex_helper.remote(df, old_other_index,
+                                                       new_index, 1)
+                                for df in other._col_partitions]
+
+        return zip(new_partitions_self, new_partitions_other)
+
+    def _inter_df_op_helper(self, func, other, axis, level):
+        if level is not None:
+            raise NotImplementedError("Mutlilevel index not yet supported "
+                                      "in Pandas on Ray")
+        axis = pd.DataFrame()._get_axis_number(axis)
+
+        # Adding two DataFrames causes an outer join.
+        if isinstance(other, DataFrame):
+            new_column_index = self.columns.join(other.columns, how="outer")
+            new_index = self.index.join(other.index, how="outer")
```

Comment:
Might be worth putting a note here (like you did in the `join` code) for the future to join on metadatas, enabling you to pass metadatas below.